ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,148 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace import tracer
+from ddtrace.constants import ERROR_MSG
+from ddtrace.constants import ERROR_TYPE
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+log = get_logger(__name__)
+
+
+def crash_flush(_, __):
+    """
+    Tags the current root span with an Impending Timeout error and
+    finishes all remaining spans.
+
+    Shutdowns the tracer at the end.
+    """
+
+    root_span = tracer.current_root_span()
+    root_span.error = 1
+    root_span.set_tag_str(ERROR_MSG, "Datadog detected an Impending Timeout")
+    root_span.set_tag_str(ERROR_TYPE, "Impending Timeout")
+
+    current_span = tracer.current_span()
+    while current_span:
+        current_span.finish()
+        current_span = current_span._parent
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        if apm_flush_deadline < 200:
+            log.debug(
+                "DD_APM_FLUSH_DEADLINE was set to %d, this value will be overridden to 200ms instead.",
+                apm_flush_deadline,
+            )
+            apm_flush_deadline = 200
+
+        remaining_time_in_millis = apm_flush_deadline
+
+    remaining_time_in_seconds = (remaining_time_in_millis - 100) / 1000
```

Comment:
If `remaining_time_in_millis` is less than apm_flush_deadline and less than 100 this value will be negative. 
Ex: when apm_flush_deadline=200 and remaining_time_in_millis=50,  `remaining_time_in_seconds = (50 - 100)/1000`. In this case we should try to finish flush all traces immediately. 

```suggestion
    remaining_time_in_seconds = max((remaining_time_in_millis - 100) / 1000, 0)
```

Reply:
This is unlikely to happen, but it would be possible if the user starts a tracer after doing some time consuming computation. So I'll commit the suggestion! 🙇🏽